kinto/plugins/quotas/listener.py

```diff
@@ -0,0 +1,235 @@
+
+    bucket_max_bytes = get_bucket_settings(settings, bucket_id, 'max_bytes')
+    bucket_max_items = get_bucket_settings(settings, bucket_id, 'max_items')
+    bucket_max_bytes_per_item = get_bucket_settings(settings, bucket_id,
+                                                    'max_bytes_per_item')
+    collection_max_bytes = get_collection_settings(settings, bucket_id,
+                                                   collection_id, 'max_bytes')
+    collection_max_items = get_collection_settings(settings, bucket_id,
+                                                   collection_id, 'max_items')
+    collection_max_bytes_per_item = get_collection_settings(
+        settings, bucket_id, collection_id, 'max_bytes_per_item')
+
+    # XXX: Maybe we want to differenciate between
+    #      bucket/collection/records/group payload.
+    max_bytes_per_item = (collection_max_bytes_per_item or
+                          bucket_max_bytes_per_item)
+
+    storage = event.request.registry.storage
+
+    if action == 'delete' and resource_name == 'bucket':
+        try:
+            storage.delete(parent_id=bucket_uri,
+                           collection_id='quota',
+                           object_id='bucket_info')
+        except RecordNotFoundError:
+            pass
+
+        collection_pattern = instance_uri(event.request, 'collection',
+                                          bucket_id=bucket_id, id='*')
+        storage.delete_all(parent_id=collection_pattern,
+                           collection_id='quota')
+        return
+
+    targets = []
+    for impacted in event.impacted_records:
+        target = impacted['new' if action != 'delete' else 'old']
+        # On POST .../records, the URI does not contain the newly created
+        # record id.
+        obj_id = target['id']
+        parts = event_uri.split('/')
+        if resource_name in parts[-1]:
+            parts.append(obj_id)
+        else:
+            # Make sure the id is correct on grouped events.
+            parts[-1] = obj_id
+        uri = '/'.join(parts)
+
+        old = impacted.get('old', {})
+        new = impacted.get('new', {})
+
+        targets.append((uri, obj_id, old, new))
+
+    try:
+        bucket_info = copy.deepcopy(
+            storage.get("quota", bucket_uri, 'bucket_info'))
+    except RecordNotFoundError:
+        bucket_info = {
+            "collection_count": 0,
+            "record_count": 0,
+            "storage_size": 0,
+        }
+
+    collection_info = {
+        "record_count": 0,
+        "storage_size": 0,
+    }
+    if collection_id:
+        try:
+            collection_info = copy.deepcopy(
+                storage.get("quota", collection_uri, 'collection_info'))
+        except RecordNotFoundError:
+            pass
+
+    # Update the bucket quotas values for each impacted record.
+    for (uri, obj_id, old, new) in targets:
+        if resource_name == "collection":
+            collection_uri = uri
```

Comment:
You are right that this might not be sufficient and we should probably get the collection_info here too.

This is to fix the batch on multiple collections for instance.